### PR TITLE
Converted to use TypeScript, Promises and automatically receive respo…

### PR DESCRIPTION
**Pull request by Tiberriver256**
Hey,

I'm not sure if you want to use this or not but thought I would share it back with you. I converted to TypeScript, it uses Promises and doesn't actually have to use files on the server or client side using this strategy:
https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.1.0/com.ibm.zos.v2r1.halu001/autosubmit.htm

### CODE REVIEW

**Comment by bushidocodes** [comment]
@Tiberriver256 - Thank you so much for the PR! My eyes are slowly recovering from a dilation at the ophthalmologist, but I'll take a look over the next day or so. I see I need to add a license on this project too. Does MIT work for you purposes?

**Comment by Tiberriver256** [comment]
Yeah that works. Let me know if you have any questions.

I really like automatically receiving the response but I think some people would still want to be able to submit a long running job and not have it tie up the FTP session waiting for it to complete. For quick jobs it's really handy though :)

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            return this.site(ftpSession, "FILEtype=JES NOJESGETBYDSN");
+          })
+          .then((responseText) => {
+            return this.get(ftpSession, remoteFileName);
+          })
+          .then ((buffer) => {
+            return resolve(buffer);
+          })
+          .catch((error) => {
+            console.error(error);
```
Comment (bushidocodes) [inline index.ts]: Perhaps we should return an explicitly rejected promise here? I could definitely imagine promise chains off this, and I think the chain might hang if we don't explicitly reject.

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+  constructor(
+    public connectionOptions: ftp.Options,
+  ) { }
+
+  /**
+   * Submits JCL to JES from a filepath, NodeJS.ReadableStream, or Buffer object and
+   * returns the result as a Promise<Buffer>
+   *
+   * @param  {string|NodeJS.ReadableStream|Buffer} input
```
Comment (bushidocodes) [inline index.ts:21]: These type annotations are really nice!

File: tests/test.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+const connectionOptions = {
+    host: "localhost",
+    password: "test",
+    user: "test",
+};
+
+const jobEntrySubsystem = new JobEntrySubsystem(connectionOptions);
+
+const JCLTestJob = `
```
Comment (bushidocodes) [inline tests/test.ts:11]: JCL actually looks pretty great in an ES6 template string, especially with your use of template tags.

**Review by bushidocodes — COMMENTED** [review]
@Tiberriver256, your code is 🔥🔥🔥🔥!!!

Really great work here. This PR massively improves things across the board. The only thing I noticed was that the promise isn't getting explicitly rejected in submitJob's catch block. I'm currently without systems access, so I haven't run your test yet, but I'll take care of that shortly.

Once we reach some consensus on the catch block thing and I successful execute your JCL, I'll approve.

File: tests/test.ts (hunk 3 above)
Comment (Tiberriver256) [reply tests/test.ts:11]: Doesn't it though :)

File: index.ts (hunk 1 above)
Comment (Tiberriver256) [reply index.ts]: Agreed good catch

**Comment by Tiberriver256** [comment]
How does that look for rejecting the promise?

**Comment by bushidocodes** [comment]
@Tiberriver256 - Looks good. I ran the test, and it seems to work as expected. I'm getting JCL errors `IEFC452I INVALID - JOB NOT RUN - JCL ERROR  361`, which I'd have to look up, but getting the job log back is good enough for me to merge. ★★★★★

**Comment by Tiberriver256** [comment]
Cool :). Might be you don't have Top Secret set up? My test was executing a CA Top Secret TSO command TSS.

**Comment by bushidocodes** [comment]
Yeah, I figured that's probably the case, but I'm not sure if JCL ERROR 361 corresponds to command not found. I'll probably replace the test JCL with something more vanilla to make it a bit more clear. Not sure if you're a pen tester / security guy given the Top Secret command, but Soldier of Fortran has some nifty things: https://github.com/mainframed

**Pull request merged**